rosidl_typesupport_cpp/test/benchmark/benchmark_type_support_dispatch.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcpputils/shared_library.hpp"
+#include "rosidl_typesupport_cpp/identifier.hpp"
+#include "rosidl_typesupport_cpp/message_type_support_dispatch.hpp"
+#include "rosidl_typesupport_cpp/service_type_support_dispatch.hpp"
+#include "rosidl_typesupport_c/type_support_map.h"
+
+#include "performance_test_fixture/performance_test_fixture.hpp"
+
+using performance_test_fixture::PerformanceTest;
+
+constexpr size_t map_size = 4u;
+constexpr const char package_name[] = "rosidl_typesupport_cpp";
+constexpr const char * identifiers[map_size] = {
+  "test_type_support1", "test_type_support2", "test_type_support3", "test_type_support4"
+};
+
+constexpr const char * symbols[map_size] = {
+  "test_message_type_support",
+  "test_message_type_support2",
+  "test_message_type_support3",
+  "test_message_type_support4"
+};
+
+rosidl_message_type_support_t get_rosidl_message_type_support(const char * identifier)
+{
+  return {identifier, nullptr, nullptr};
+}
+
+rosidl_service_type_support_t get_rosidl_service_type_support(const char * identifier)
+{
+  return {identifier, nullptr, nullptr};
+}
+
+type_support_map_t get_typesupport_map(void ** library_array)
+{
+  return type_support_map_t{
+    map_size,
+    package_name,
+    identifiers,
+    symbols,
+    library_array,
+  };
+}
+
+BENCHMARK_F(PerformanceTest, message_typesupport_handle_function)(benchmark::State & st)
+{
+  rosidl_message_type_support_t type_support_cpp_identifier =
+    get_rosidl_message_type_support(rosidl_typesupport_cpp::typesupport_identifier);
+  rcpputils::SharedLibrary * library_array[map_size] = {nullptr, nullptr, nullptr};
+  type_support_map_t support_map = get_typesupport_map(reinterpret_cast<void **>(&library_array));
+  type_support_cpp_identifier.data = &support_map;
+
+  reset_heap_counters();
+
+  for (auto _ : st) {
+    // Successfully load library and find symbols
+    auto * result = rosidl_typesupport_cpp::get_message_typesupport_handle_function(
+      &type_support_cpp_identifier,
+      "test_type_support1");
+    if (nullptr == result) {
+      st.SkipWithError("rosidl_typesupport_cpp::get_message_typesupport_handle_function failed");
+    }
+  }
```

Comment:
These tests have the same problem as I described in the `rmw_implementation` PR. The first call loads the library, and the subsequent calls don't need to.

You can see the effect on the results, where the time is really small, and the `heap_allocations` aren't consistent, like you would expect (I'm seeing 2.00002 allocations on average).
```suggestion
    // Unload for the next iteration
    for (size_t i = 0; i < map_size; i++) {
      if (library_array[i] != nullptr) {
        delete library_array[i];
        library_array[i] = nullptr;
      }
    }
  }
```

If you unload the library like this between iterations, each iteration has to re-load the library. The results are significant. Without unloading, the average duration is around 600 ns, and performs an average of ~2 heap allocations per iteration. If you force each iteration to load the libraries, the duration shoots up to around 38000 ns on my machine, and each iteration performs exactly 30 heap allocations.

If you ever see a non-integer for the `heap_allocations` count, it means that there is a different amount of stuff happening during each iteration of the timing loop. Sometimes that is expected, but for microbenchmarks like this, it's almost always an indicator that you aren't resetting the state between iterations.

All 4 of the tests in this PR will need this snippet added. I know we don't really care to include the snip in our timing results, but getting consistent results is far more important here.

If you're also interested in timing the cache-hit case, that would also be interesting. Just take what you have and perform an initial invocation outside of the loop and before calling `reset_heap_counters` to prime the cache, so that every one of the invocations inside of the timing loop are a cache hit.

Reply:
Thank you @cottsay for clarifying me this. It was not clear for me the float heap allocations